### ISSUE TITLE
New Release

### ISSUE THREAD

**Issue by svozza** [caolan/highland#267]
There's been a lot of new features that have been added since the last release, it might be best to pull the trigger on this sooner rather than later. Here's a list of all the changes, I think we should copy and paste it into the release notes in the [Releases] (https://github.com/caolan/highland/releases) section.

### New additions
* `drop`: Ignores the first `n` values of a stream and then emits the rest. #75 #244 
* `done`: Calls the supplied function once the stream has ended. #161
* `sort`: Collects all values together then emits each value individually but in sorted order. #169 #245
* `streamifyAll`: Takes an object or a constructor function and returns that object or constructor with streamified versions of its function properties. #226 
* `Iterator` Support: ECMA2015 (aka ES6) style iterators can now be passed to the Highland constructor function. #235 
* `slice`: Creates a new stream with the values from the source in the range of specified in the`start` and `end` parameters. #250 
* `batchWithTimeOrCount`: Takes one Stream and batches incoming data within a maximum time frame into arrays of a maximum length. #284

### Improvements
* `each` now returns an empty stream rather than nothing. #161.
* Ensure `through` propagates Node stream errors. #240 
* Preserve `this` context of wrapped function when using `wrapCallback`. #248 
* Update `tranduce` to use latest version of  [transformer protocol](https://github.com/cognitect-labs/transducers-js#transformer-protocol). #261

**Comment by vqvu** [comment]
+1 on new release.

One more improvement

* `each` now returns an empty stream rather than nothing. #161.

**Comment by svozza** [comment]
Ah I'd missed that one. Incidentally, if we're finally bringing in `drop` should we also merge `slice` (#250)?

**Comment by jeromew** [comment]
I tried to make a release but eslint is not happy when calling `npm test`

lib/index.js
  2896:40  error  err is already declared in the upper scope   no-shadow
  3281:29  error  memo is already declared in the upper scope  no-shadow
  3291:27  error  memo is already declared in the upper scope  no-shadow
  3987:18  error  Empty block statement                        no-empty

✖ 4 problems (4 errors, 0 warnings)

do you have the same issue on your side ? it is probably due to recent  commits.

**Comment by vqvu** [comment]
Slice has optional arguments. Not sure what to do about that...

@jeromew, that's strange. I don't see it. Fresh clone and everything.

**Comment by quarterto** [comment]
It's possible to configure Travis to [release to npm when a tagged commit passes](http://docs.travis-ci.com/user/deployment/npm/). Do we want to consider this so we don't have to wait for @caolan or @jeromew to release?

**Comment by vqvu** [comment]
Wouldn't this expose their NPM credentials to the public? Unless `travis
encrypt` uses a repository-specific public key.

**Comment by svozza** [comment]
Regarding `slice`, I had noticed that although it's worth noting that it actually doesn't affect the currying. These all work as expected:

```javascript
var arr = [1, 2, 3, 4, 5];

var first = _.slice(1);
var second = _.slice(1, 4);

first(3, _(arr)).toArray(_.log);
//=> [2, 3]

second(_(arr)).toArray(_.log);
//=> [2, 3, 4]
```
I prefer fixed arity functions though so a solution would be to just remove the default values in slice and be explicit in the implementation of `drop`:

```javascript
Stream.prototype.drop = function (n) {
    if (n <= 0) {
        return this;
    }
    return this.slice(n, Infinity);
};
exposeMethod('drop');
```

**Comment by vqvu** [comment]
The problem with optional arguments has always been if people try to do

```javascript
var drop5 = _.slice(5);
drop(_([1, 2, 3, 4, 5])).toArray.log(_.log);
// => an error.
```

Not having optional args means people can't accidentally mess up like this. But, I'm also OK with relaxing our stance on optional args if we put a warning in big red letters in the docs that says the top-level `slice` requires an explicit end parameter. We'll just be pushing the hard work to the user.

**Comment by svozza** [comment]
I think I would rather keep things consistent and just have the blanket rule that transforms don't take optional arguments, they complicate the API and once that genie is out of the bottle it's hard put it back in.

**Comment by vqvu** [comment]
I don't feel strongly either way, but `slice` is so ubiquitous that not having `end` be optional could cause the same kinds of confusion. Or we could implement `drop` and `take` in terms of a `slice`-like function, but not expose a `slice` transform.

**Comment by svozza** [comment]
Sorry, meant to reply to this. `slice` is quite useful though so it would be a shame to lose it. Does anyone else have any input? I'll go along with the consensus.

**Comment by vqvu** [comment]
We could go the `reduce` route and have a `slice1`.

On the other hand, `slice(start, end)` is equivalent to `.drop(start).take(end - start)`.

**Comment by caolan** [comment]
I'd vote for keeping fixed arity, I understand this would likely trip people up in the case of slice though. Any more suggestions on alternative names for slice? Perhaps 'substream(start, end)', similar to substr()?

**Comment by jeromew** [comment]
@vqvu regarding the release, I still have the issue with linting and don't understand why you don't see it in a fresh clone

for instance, the warning

    2896:40 error err is already declared in the upper scope no-shadow

correctly detects a linting problem when you look at line 2896 because a closure shadows the `err` param that is already used in the scope.

I am surprised to be the only one with this problem.

**Comment by jeromew** [comment]
@vqvu ok ; after removing node_modules altogether + a fresh `npm install`, the linting issue has disappeared.

I still don't understand it because it looks like the 

    2896:40 error err is already declared in the upper scope no-shadow

issue really exists in the code line 2896 (there is an `err` variable used by the closure that shadows the `err` from the upper scope)

am i misunderstanding this 'no-shadow' directive ?

**Comment by jeromew** [comment]
@svozza you said

    I think we should copy and paste it into the release notes in the Releases section.

I think we should use the CHANGELOG file instead. It hasn't been used since 2.0.0 but should probably be used from now on.

regarding other options for `slice`, there could be `cut`, `extract`, `section`. `cut` is the word used for video cutting for instance when you want to extract a slice of a video.

**Comment by svozza** [comment]
Yeah, I just mentioned the Releases section because I saw it on the Bluebird page but changelog is as good a place as any. If we change the file name to changelog.md we can get the markup to work in it as well.

**Comment by vqvu** [comment]
We'll want to fix #205 before we release or the page that gets generated will be broken again.

@jeromew Perhaps the `no-shadow` was turned off by default in a later version of eslint? We don't explicitly enable that check in our `.eslintrc`.

I like `substream` since `cut

**Comment by vqvu** [comment]
Thought about it some more, and I now think it's just fine to have `slice` without the optional argument (so objection withdrawn! :)).

After all, if I was OK with `slice` and `slice1`, then I should be OK with `slice` and `drop`.

The PR is fine to merge, I think, after applying @svozza's proposed fix.

**Comment by jeromew** [comment]
@svozza, @vqvu do you want to release with or without a merge of a renamed `slice` ? what do you think of the different options proposed & do you have other ideas ?

**Comment by jeromew** [comment]
@vqvu so if I understand well you had the main objection regarding the `slice` name but that objection is gone now. @caolan was your message an objection also or do you agree with the `slice` name ?

**Comment by vqvu** [comment]
@jeromew Yes.

**Comment by svozza** [comment]
I'd say everything for this release is merged now, we should depoly to NPM as soon as possible.

**Comment by jeromew** [comment]
Sorry everyone about the delay.
I updated the `CHANGELOG`, renamed it `CHANGELOG.md`, mentioned `batchWithTimeOrCount` that was merged in between and pushed 2.5.0 to npm.
thanks to everyone for this release !
